src/materialized/src/prof.rs

```diff
@@ -0,0 +1,122 @@
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::{ffi::CString, time::Instant};
+use tempfile::NamedTempFile;
+
+#[derive(Copy, Clone, Debug)]
+pub enum ProfStartTime {
+    Instant(Instant),
+    TimeImmemorial,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct JemallocProfMetadata {
+    pub start_time: Option<ProfStartTime>,
+}
+
+#[derive(Debug)]
+// Per-process singleton object allowing control of jemalloc profiling facilities.
+pub struct JemallocProfCtl {
+    md: JemallocProfMetadata,
+}
+
+#[cfg(target_os = "macos")]
+impl JemallocProfCtl {
+    fn get() -> Option<Self> {
+        None
+    }
+    pub fn get_md(&self) -> JemallocProfMetadata {
+        unreachable!()
+    }
+
+    pub fn activate(&mut self) -> Result<(), jemalloc_ctl::Error> {
+        unreachable!()
+    }
+
+    pub fn deactivate(&mut self) -> Result<(), jemalloc_ctl::Error> {
+        unreachable!()
+    }
+
+    pub fn dump(&mut self) -> anyhow::Result<std::fs::File> {
+        unreachable!()
+    }
+}
+
+#[cfg(not(target_os = "macos"))]
+impl JemallocProfCtl {
+    // Creates and returns the global singleton.
+    fn get() -> Option<Self> {
+        // SAFETY: "opt.prof" is documented as being readable and returning a bool:
+        // http://jemalloc.net/jemalloc.3.html#opt.prof
+        let prof_enabled: bool = unsafe { raw::read(b"opt.prof\0") }.unwrap();
+        if prof_enabled {
+            // SAFETY: "opt.prof_active" is documented as being readable and returning a bool:
+            // http://jemalloc.net/jemalloc.3.html#opt.prof_active
+            let prof_active: bool = unsafe { raw::read(b"opt.prof_active\0") }.unwrap();
+            let start_time = if prof_active {
+                Some(ProfStartTime::TimeImmemorial)
+            } else {
+                None
+            };
+            let md = JemallocProfMetadata { start_time };
+            Some(Self { md })
+        } else {
+            None
+        }
+    }
+
+    pub fn get_md(&self) -> JemallocProfMetadata {
+        self.md
+    }
+
+    pub fn activate(&mut self) -> Result<(), jemalloc_ctl::Error> {
+        // SAFETY: "prof.active" is documented as being readable and returning a bool:
+        // http://jemalloc.net/jemalloc.3.html#prof.active
+        unsafe { raw::write(b"prof.active\0", true) }?;
+        if self.md.start_time.is_none() {
+            self.md.start_time = Some(ProfStartTime::Instant(Instant::now()));
+        }
+        Ok(())
+    }
+
+    pub fn deactivate(&mut self) -> Result<(), jemalloc_ctl::Error> {
+        // SAFETY: "prof.active" is documented as being readable and returning a bool:
+        // http://jemalloc.net/jemalloc.3.html#prof.active
+        unsafe { raw::write(b"prof.active\0", false) }?;
+        self.md.start_time = None;
+        Ok(())
+    }
+
+    pub fn dump(&mut self) -> anyhow::Result<std::fs::File> {
+        let f = NamedTempFile::new()?;
+        let path = CString::new(f.path().as_os_str().as_bytes().to_vec()).unwrap();
```

Comment:
This doesn't matter, but I believe you could use `CStr` instead of `CString` and not need the `to_vec` call.

Reply:
no, it needs to do an allocation to stick the zero byte at the end.

Reply:
oh right.